Truncate tlog cli should assign global checkpoint

dnhatn opened the pull request:
We are targeting to always have a safe index once the recovery is done. This invariant does not hold if the translog is manually truncated by users because the truncate translog cli resets the global checkpoint to unassigned. This commit assigns the global checkpoint to the max_seqno  of the last commit when truncating translog. We can only safely do it because the truncate translog command will generate a new history uuid for that shard. With a new history UUID, sequence-based recovery between that shard and other old shards will be disabled.

Relates #28181

dnhatn commented:
> Recovery from local shards and snapshots can set the global checkpoint to the local checkpoint as this is the only copy of the shard.

The invariant does not hold for recovering from local shards and snapshots. I think we can assign the global checkpoint to max_seqno in these cases.

bleskes commented on server/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java:
```diff
@@ -135,6 +135,13 @@ protected void execute(Terminal terminal, OptionSet options, Environment env) th
                 String translogGeneration = commitData.get(Translog.TRANSLOG_GENERATION_KEY);
                 String translogUUID = commitData.get(Translog.TRANSLOG_UUID_KEY);
+                final long globalCheckpoint;
+                // In order to have a safe commit invariant, we have to assign max_seqno of the last commit to the global checkpoint.
```
I think we need to do more here to get to a healthy shard - we need to also move the local checkpoint to the max seq # . Otherwise we have the weird state where local checkpoint < global checkpoint. By setting the local checkpoint to the max seq# we declare all missing ops (that were in the translog) as no ops.

bleskes replied:
I would also like to double check why no test failed due to the above. We should get a tight grip on when our invariants (gcp <= lcp <= max seq#) kick in.

dnhatn commented:
@bleskes, I've updated the description and moved the local checkpoint to max_seqno in the truncate translog cli. Could you please take another look.

No test failed because we execute `fillSeqNoGaps` when recovering from the store. This makes the local checkpoint be in sync the global checkpoint and the max_seqno. The local checkpoint from the last commit can be lower than the global checkpoint if the global checkpoint advanced (eg. the local checkpoint also advanced) but we do not flush.

bleskes commented on server/src/test/java/org/elasticsearch/index/translog/TruncateTranslogIT.java:
```diff
@@ -214,6 +217,9 @@ public void testCorruptTranslogTruncation() throws Exception {
         assertThat(replicaRecoveryState.getIndex().toString(), replicaRecoveryState.getIndex().recoveredFileCount(), greaterThan(0));
+        // Ensure that the global checkpoint is restored from the max seqno of the last commit.
+        final SeqNoStats seqNoStats = getSeqNoStats("test", 0);
+        assertThat(seqNoStats.getGlobalCheckpoint(), equalTo(seqNoStats.getMaxSeqNo()));
```
also check the local checkpoint?

dnhatn commented:
Thanks @bleskes for reviewing.
